### DISCUSSION TITLE
Filament for game dev

### DISCUSSION THREAD

**Discussion by MicahMartin** [google/filament#4761] [Q&A] [answered]
Hey guys, is filament suitable for game dev? I was wondering if I could use it as a 3d rendering API for a custom game engine

**Comment by romainguy** [comment, marked as answer]
There’s one game on Steam using it: https://store.steampowered.com/app/1505010/Blockdown/

There are features missing that may be a problem for some types of games, like local light probes and instancing.

**Reply by prideout** [reply]
For certain types of games, Filament works really well.  The aforementioned game achieves "instancing" for tiles as follows.

On every frame, it updates a small (100x26) floating-point texture that has two layers (positions and orientations).  The values in this texture are used to inform a custom vertex material.

The entire set of tiles is rendered in a single draw call using an immutable vertex buffer. The vertex buffer contains 2600 copies of the same tile, and all copies co-exist at the world origin. Each copy is tagged with a custom vertex attribute that tells the vertex shader the tile id.

Animation and variation is achieved via the geotexture, not the vertex buffer. This is efficient because each tile is comprised of many vertices, and all tiles have the same geometry.

**Reply by MicahMartin** [reply]
Wow, that's super interesting. Do you think filament work for a fighting game engine?
